src/integer/mat_z/tensor.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright © 2023 Marvin Beckmann
+//
+// This file is part of qFALL-math.
+//
+// qFALL-math is free software: you can redistribute it and/or modify it under
+// the terms of the Mozilla Public License Version 2.0 as published by the
+// Mozilla Foundation. See <https://mozilla.org/en-US/MPL/2.0/>.
+
+//! This module contains the implementation of the `tensor` product.
+
+use super::MatZ;
+use crate::traits::{GetNumColumns, GetNumRows, Tensor};
+use flint_sys::{
+    fmpz::{fmpz_is_zero, fmpz_mul},
+    fmpz_mat::fmpz_mat_entry,
+};
+
+impl Tensor for MatZ {
+    /// Computes the tensor product of `self` with `other`
+    ///
+    /// Parameters:
+    /// - `other`: the value with which the tensor product is computed.
+    ///
+    /// Returns the tensor product of `self` with `other`.
+    ///
+    /// # Example
+    /// ```
+    /// use qfall_math::integer::MatZ;
+    /// use qfall_math::traits::Tensor;
+    /// use std::str::FromStr;
+    ///
+    /// let mat_1 = MatZ::from_str("[[1, 0, 0],[0, 1, 0],[0, 0, 1]]").unwrap();
+    /// let mat_2 = MatZ::from_str("[[1, 2, 1],[3, 4, 1]]").unwrap();
+    ///
+    /// let mat_3 = mat_1.tensor(&mat_2);
+    /// ```
+    fn tensor(&self, other: &Self) -> Self {
+        let columns_other = other.get_num_columns();
+        let rows_other = other.get_num_rows();
+
+        // we can unwrap since we know that the dimensions are positive
+        let mut out = MatZ::new(
+            self.get_num_rows() * rows_other,
+            self.get_num_columns() * columns_other,
+        )
+        .unwrap();
+
+        for i in 0..self.get_num_rows() {
+            for j in 0..self.get_num_columns() {
+                let entry = unsafe { fmpz_mat_entry(&self.matrix, i, j) };
+
+                if unsafe { 1 != fmpz_is_zero(entry) } {
+                    for i_other in 0..rows_other {
+                        for j_other in 0..columns_other {
+                            unsafe {
+                                fmpz_mul(
+                                    fmpz_mat_entry(
+                                        &out.matrix,
+                                        i * rows_other + i_other,
+                                        j * columns_other + j_other,
+                                    ),
+                                    entry,
+                                    fmpz_mat_entry(&other.matrix, i_other, j_other),
+                                )
+                            }
+                        }
+                    }
+                }
```

Comment:
please outsource this part:
- the nicest way would be via two efficient functions in the fmpz_helpers.rs file in src/integer/z/: one for concatenation of matrices and one for scalar multiplication (probably not as efficient as your current implementation without having several trade offs)
- just one function specifically for this.

It's just too many for-loops in each other/ tabs in one line.

Reply:
I prefer the second version, since it is probably more efficient. While programming I considered your first idea, but due to efficiency scraped that idea

Reply:
I will add a subroutine for the latter